compiler/rustc_middle/src/query/mod.rs

```diff
@@ -768,6 +767,28 @@ rustc_queries! {
         desc { |tcx| "comparing impl items against trait for `{}`", tcx.def_path_str(impl_id) }
     }
 
+    /// Given a trait `fn_def_id`, create and return the associated items list for that function.
+    query assoc_items_for_rpitits(fn_def_id: DefId) -> &'tcx [DefId] {
+        desc { |tcx| "creating associated items for rpits returned by `{}`", tcx.def_path_str(fn_def_id) }
+        cache_on_disk_if { fn_def_id.is_local() }
+        separate_provide_extern
+    }
+
+    /// Given an impl fn `impl_fn_def_id` create and return all the impl associated items corresponding to the rpits on the corresponding trait fn.
+    query impl_assoc_items_for_rpitits(impl_fn_def_id: DefId) -> &'tcx [DefId] {
```

Comment:
Why does this need to be two queries? Shouldn't we be able to see if the `fn_def_id` comes from a trait or an impl, and dispatch to two sets of logic accordingly?